internal/querynodev2/delegator/delegator.go

```diff
@@ -203,6 +207,140 @@ func (sd *shardDelegator) modifyQueryRequest(req *querypb.QueryRequest, scope qu
 	return nodeReq
 }
 
+// OptimizeSearchBasedOnClustering optimize SearchRequest based on segments' clustering info.
+// Basic rule: calculate distance between search vector and clustering centroid of each segment,
+// only search on top x% nearest segment
+// Todo support call external hook
+func (sd *shardDelegator) OptimizeSearchBasedOnClustering(req *querypb.SearchRequest, sealeds []SnapshotItem) (*querypb.SearchRequest, []SnapshotItem) {
+	if !paramtable.Get().QueryNodeCfg.EnableSearchBasedOnClustering.GetAsBool() ||
+		!req.GetReq().GetClusteringOptions().GetEnable() {
+		log.Debug("skip optimize based on clustering info")
+		return req, sealeds
+	}
+
+	// basic optimize rule
+	topK := req.GetReq().GetTopk()
+	filterRatio := req.GetReq().GetClusteringOptions().GetFilterRate()
+	dim := req.GetReq().GetDim()
+	metricType := req.GetReq().GetMetricType()
+	var phg commonpb.PlaceholderGroup
+	err := proto.Unmarshal(req.GetReq().GetPlaceholderGroup(), &phg)
+	if err != nil {
+		log.Warn("fail to parse SearchRequest PlaceholderGroup", zap.Error(err))
+		return req, sealeds
+	}
+	log.Debug("optimizeSearchBasedOnClustering",
+		zap.String("metricType", metricType),
+		zap.Int32("dim", dim),
+		zap.Int("length", len(phg.GetPlaceholders())),
+		zap.Any("phg", phg))
+
+	vectorsBytes := phg.GetPlaceholders()[0].GetValues()
+	searchVectors := make([][]float32, len(vectorsBytes))
+	for i, vectorBytes := range vectorsBytes {
+		searchVectors[i] = clustering.DeserializeFloatVector(vectorBytes)
+	}
+
+	var segments = make([]SegmentEntry, 0)
+	for _, sealed := range sealeds {
+		segments = append(segments, sealed.Segments...)
+	}
+
+	type segmentDistanceStruct struct {
+		segment  SegmentEntry
+		distance float32
+	}
+
+	var vectorSegmentDistances = make([][]segmentDistanceStruct, 0)
+	for _, searchVector := range searchVectors {
+		var vectorSegmentDistance = make([]segmentDistanceStruct, 0)
+		for _, segment := range segments {
+
+			if segment.ClusteringInfos != nil {
+				for _, clusteringInfo := range segment.ClusteringInfos {
```

Comment:
now we do not handle the multiple clustering groups case, can do it later

Reply:
yes